include/sparrow/buffer_adaptor.hpp

```diff
@@ -0,0 +1,503 @@
+    {
+        return const_iterator(data() + size());
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::const_iterator buffer_adaptor<To, From>::cbegin() const noexcept
+    {
+        return begin();
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::const_iterator buffer_adaptor<To, From>::cend() const noexcept
+    {
+        return end();
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::reverse_iterator buffer_adaptor<To, From>::rbegin() noexcept
+    {
+        return reverse_iterator(end());
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::reverse_iterator buffer_adaptor<To, From>::rend() noexcept
+    {
+        return reverse_iterator(begin());
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::const_reverse_iterator
+    buffer_adaptor<To, From>::rbegin() const noexcept
+    {
+        return const_reverse_iterator(end());
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::const_reverse_iterator
+    buffer_adaptor<To, From>::rend() const noexcept
+    {
+        return const_reverse_iterator(begin());
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::const_reverse_iterator
+    buffer_adaptor<To, From>::crbegin() const noexcept
+    {
+        return rbegin();
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::const_reverse_iterator
+    buffer_adaptor<To, From>::crend() const noexcept
+    {
+        return rend();
+    }
+
+    // Capacity
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr buffer_adaptor<To, From>::size_type buffer_adaptor<To, From>::size() const noexcept
+    {
+        const double new_size = static_cast<double>(m_buffer.size()) * m_from_to_size_ratio;
+        SPARROW_ASSERT(
+            std::trunc(new_size) == new_size,
+            "The size of the buffer is not a multiple of the size of the new type"
+        );
+        return static_cast<size_type>(new_size);
+    }
+
+    template <typename To, typename From>
+        requires(sizeof(From) < sizeof(To))
+    constexpr typename buffer_adaptor<To, From>::size_type buffer_adaptor<To, From>::max_size() const noexcept
+    {
+        return static_cast<size_type>(static_cast<double>(m_buffer.max_size()) * m_from_to_size_ratio);
```

Comment:
Could you isolate/factorize that calculation so that it's not repeated in `capacity()` etc.? From what I understand only the initial value is a paremeter.

Reply:
I modified the code to avoid this long line of code